### PR TITLE
feat(helm): add requirements.yaml support

### LINKED ISSUE #874: Feature: Add dependency spec to Chart.yaml (was Kubernetesfile)
## what
* Introduce the concept of a `Kubernetesfile` that can be included with applications to express how the service is deployed. It would contain the YAML or JSON of kubernetes resources.
* Support remote dependencies expressed via a `git` URL to a repo containing a top-level `Kubernetesfile`
 
## why

As applications change or evolve, the way they are instantiated may change. For example, new environment variables may be introduced or removed, new command line arguments may be accepted, additional ports exposed, etc. From a maintenance perspective, it would be convenient if the author of the app could keep the `Kubernetesfile` updated as the code evolves rather than try to keep multiple repositories (e.g. external Helm Charts) current. 

This is similar to what the `Dockerfile` expresses for building an application, the `Jenkinsfile` for integrating it with CI/CD pipelines and (once-upon-a-time) the `Vagrantfile` for setting up a local environment. 

## our use-case
We have a product that is composed of a dozen proprietary micro-services as well as other off-the-shelf components like statsd.  We want to use a repository that maps out the `Chart` for how to deploy our product in our Kubernetes cluster (the way it works now), but we want service owners to "own" the definitions of how to deploy their service much the way they "own" their `Dockerfiles` for building them. 

We like the way it works now, but would like this alternative approach for including services that we own.

### PR DESCRIPTION
**Pull request by technosophos**
This adds...
- support for a `$CHART/requirements.yaml` file, along with a lock file
- `helm dependency list|build|update`
- better testing for repository servers

This supersedes #1212.

Closes #874

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/helm/1221)

### CODE REVIEW

File: cmd/helm/helmpath/helmhome.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+*/
+
+package helmpath
+
+import (
+	"fmt"
+	"path/filepath"
+)
+
+// HelmHome describes the location of a CLI configuration.
```
Comment (adamreese) [inline cmd/helm/helmpath/helmhome.go]: Doc string should be `Home ...`

File: cmd/helm/dependency.go [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+			}
+		}
+		if !found {
+			fmt.Fprintf(l.out, "WARNING: %q is not in requirements.yaml.\n", f)
+		}
+	}
+
+}
+
+func lockpath(chartpath string) string {
```
Comment (adamreese) [inline cmd/helm/dependency.go]: deadcode?

File: cmd/helm/helmpath/helmhome.go (hunk 1 above)
Comment (technosophos) [inline cmd/helm/helmpath/helmhome.go]: Fixed

File: cmd/helm/downloader/manager.go [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+	if err != nil {
+		return false
+	}
+	return au.String() == bu.String()
+}
+
+// findChartURL searches the cache of repo data for a chart that has the name and the repourl specified.
+//
+// In this current version, name is of the form 'foo-1.2.3'. This will change when
+// the repository index stucture changes.
```
Comment (adamreese) [inline cmd/helm/downloader/manager.go]: almost like a index structure

File: cmd/helm/dependency_build.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	f.StringVar(&dbc.keyring, "keyring", defaultKeyring(), "The keyring containing public keys.")
+
+	return cmd
+}
+
+func (d *dependencyBuildCmd) run() error {
+	man := &downloader.Manager{
+		Out:       d.out,
+		ChartPath: d.chartpath,
+		HelmHome:  helmpath.Home(d.helmhome),
```
Comment (adamreese) [inline cmd/helm/dependency_build.go]: already a `helmpath.Home`

File: cmd/helm/dependency.go (hunk 2 above)
Comment (technosophos) [inline cmd/helm/dependency.go]: Fixed

File: cmd/helm/dependency_build.go (hunk 4 above)
Comment (technosophos) [inline cmd/helm/dependency_build.go]: Good catch. Fixed.

**Pull request merged**